File: D1L1/LuckyRemainder.java
public class LuckyRemainder {
    public int pow2mod9(int n) {
        switch (n % 6) {
            case 0 : return 1;
            case 1 : return 2;
            case 2 : return 4;
            case 3 : return 8;
            case 4 : return 7;
            case 5 : return 5;
        }
        return 0;
    }
    public int getLuckyRemainder(String X) {
        int sum = 0;
        for (int i = 0; i < X.length(); ++i) {
            sum += X.charAt(i) - '0';
        }
        sum %= 9;
        return (sum * pow2mod9(X.length() - 1)) % 9;
    }
}

File: D1L1/LuckyRemainder.hs
module LuckyRemainder where
import Data.Char

pow2mod9 :: Int -> Int
pow2mod9 0 = 1
pow2mod9 1 = 2
pow2mod9 2 = 4
pow2mod9 3 = 8
pow2mod9 4 = 7
pow2mod9 5 = 5
pow2mod9 n = pow2mod9 $ mod n 6

luckyRemainder :: [Int] -> Int
luckyRemainder ds = mod (mod (sum ds) 9 * pow2mod9 (length ds - 1)) 9

getLuckyRemainder :: String -> Int
getLuckyRemainder = luckyRemainder . map toDigit
    where toDigit c = ord c - ord '0'
